(docs) add a page about the OCaml Infrastructure (#3107)
ocaml/ocaml.org: data/tool_pages/platform/2in_00_the_ocaml_infrastructure.md

---
id: "ocaml-infrastructure"
short_title: "The OCaml Infrastructure"
title: "The OCaml Infrastructure: Tools and Services for the OCaml Ecosystem"
description: "An overview of the services that belong to the OCaml Infrastructure."
category: "OCaml Infrastructure"
---

The OCaml ecosystem is supported by a [robust infrastructure built around OCurrent](https://github.com/ocurrent/overview). The OCaml Infrastructure powers many essential services that OCaml developers rely on daily.

## What is OCurrent?

[OCurrent](https://ocurrent.org) is an OCaml eDSL for creating processing pipelines that automatically adjust when their inputs change. For example, when a Git repository is updated, OCurrent pipelines can automatically rebuild, test, and deploy the changes without manual intervention.

## OCaml Infrastructure Services

The OCaml infrastructure services can be grouped into three main categories: Community infrastructure services, services for individual projects, and infrastructure operation services. Here we present a brief overview of the major services from the first two categories.

### Services for Individual Projects

> These services used for your own projects to help with reproducible builds, testing, benchmarking, and continuous integration:

Reviewer: ```suggestion
These services can be used for your own projects to help with reproducible builds, testing, benchmarking, and continuous integration:
```

#### Docker Base Images

**Service:** [docker-base-images](https://github.com/ocurrent/docker-base-images)

> **Website:** [images.ci.ocaml.org](https://images.ci.ocaml.org)

Reviewer: I wonder if this should be something like

```suggestion
**Source:** [docker-base-images](https://github.com/ocurrent/docker-base-images)

**Service:** [images.ci.ocaml.org](https://images.ci.ocaml.org)
```

or maybe "Repo"/"Service" or somtehing? I'm thinking this because the "Website" is actually a view into the running service, but the github repo is just the source code from which the service is deployed.

> This service builds the official `ocaml/opam` Docker images for various Linux distributions, OCaml versions, compiler flags, and architectures (including x86, ARM, PowerPC, s390x, and RISC-V). These images provide a consistent environment for development and testing and are used by many other CI services.

Reviewer: Since I think it is nice to link the first time something is mentioned:

```suggestion
This service builds the official [`ocaml/opam`](https://hub.docker.com/r/ocaml/opam) Docker images for various Linux distributions, OCaml versions, compiler flags, and architectures (including x86, ARM, PowerPC, s390x, and RISC-V). These images provide a consistent environment for development and testing and are used by many other CI services.
```

> **Using the pre-built OCaml Docker images:** You can view the available [pre-built Docker images for various OCaml configurations at DockerHub](https://hub.docker.com/r/ocaml/opam).

Reviewer: If we are linking the main docker hub page when first mentioned above, maybe it is then helpful to have this go to 

```suggestion
**Using the pre-built OCaml Docker images:** You can view the available [pre-built Docker images for various OCaml configurations at DockerHub](https://hub.docker.com/r/ocaml/opam/tags).
```

Which presents people with the list of platforms that is the topic here?

#### OCaml CI

**Service:** [ocaml-ci](https://github.com/ocurrent/ocaml-ci)

**Website:** [ocaml.ci.dev](https://ocaml.ci.dev)

A CI service for OCaml projects hosted on GitHub. It automatically tests projects against multiple OCaml versions and OS platforms by examining your opam files to determine build requirements.

**How to enable OCaml CI on your GitHub project:** Check out the ["Getting Started" guide on the OCaml CI website](https://ocaml.ci.dev/getting-started).

#### Continuous Benchmarking

**Service:** [current-bench](https://github.com/ocurrent/current-bench)

**Website:** [bench.ci.dev](https://bench.ci.dev)

Provides continuous benchmarking for OCaml projects to track performance across different commits and branches.

> **How to enable current-bench on your GitHub project:** Check out the [current-bench User Manual](https://github.com/ocurrent/current-bench/blob/main/doc/user_manual.md).

Reviewer: I don't know that we should be listing and advertising this. The readme of current-bench says:

> current-bench doesn't accept enrollment for community projects at this time, but feel free to [open an issue](https://github.com/ocurrent/current-bench/issues/new) to tell us you're interested.

At least, this is not actually something that we are inviting people to use in their personal projects?

Author: Ah ok, this wasn't obvious to me at all, from the overview page, and from the way we've been giving people public updates about upgrades happening on the infrastructure.

Author: I removed all the benchmark / CI services

#### OCaml-Multicore-CI

**Service:** [ocaml-multicore-ci](https://github.com/ocurrent/ocaml-multicore-ci)

**Website:** [ocaml-multicore.ci.dev](https://ocaml-multicore.ci.dev)

Tests OCaml projects against multicore OCaml versions - particularly useful if you want to ensure your code is compatible with OCaml 5.x.

> **How to enable OCaml-Multicore-CI on your GitHub project:** Check out the [OCaml-Multicore-CI README on GitHub](https://github.com/ocurrent/ocaml-multicore-ci).

Reviewer: I have not been involved with this, but I don't know what its use case is beyond ocaml-ci now that 5 is out for some time and we are running ocaml 5.x in ocaml-ci.

It doesn't seem to me that this is a service we want people thinking is maintained and available for their use on individual projects.

---
id: "ocaml-infrastructure"
short_title: "The OCaml Infrastructure"
title: "The OCaml Infrastructure: Tools and Services for the OCaml Ecosystem"
description: "An overview of the services that belong to the OCaml Infrastructure."
category: "OCaml Infrastructure"
---

The OCaml ecosystem is supported by a [robust infrastructure built around OCurrent](https://github.com/ocurrent/overview). The OCaml Infrastructure powers many essential services that OCaml developers rely on daily.

## What is OCurrent?

[OCurrent](https://ocurrent.org) is an OCaml eDSL for creating processing pipelines that automatically adjust when their inputs change. For example, when a Git repository is updated, OCurrent pipelines can automatically rebuild, test, and deploy the changes without manual intervention.

## OCaml Infrastructure Services

> The OCaml infrastructure services can be grouped into three main categories: Community infrastructure services, services for individual projects, and infrastructure operation services. Here we present a brief overview of the major services from the first two categories.

Reviewer: Looks like this needs an update too? Maybe to just say something like

```suggestion
The OCaml infrastructure provides services to support a sustainable and well integrated software ecosystem. Many of these services can also be used directly by individual and organizations in their own projects. Here we describe the main services you may encounter in operation or make use of in your projects.
```

#### Docker Base Images

**Source:** [docker-base-images](https://github.com/ocurrent/docker-base-images)

> **Service:** [images.ci.ocaml.org](https://images.ci.ocaml.org)

Reviewer: The other correlate entries below use `Website`. Either one is fine, IMO, but I guess they may as well be consistent?

Author: resolved by marking "Source Code" vs "Website"
